quine-core/src/main/scala/com/thatdot/quine/graph/NodeActor.scala

```diff
@@ -139,6 +134,73 @@ private[graph] class NodeActor(
     case MergedHere(other) => !mergedIntoHere.contains(other)
   }
 
+  /**
+    * Drop duplicate events, and wrap the Events being kept in an `EventTime`.
+    *
+    * An event is considered a duplicate (and is dropped) if:
+    *  - the same event appears later in the sequence of events, or
+    *  - the event would have no effect when applied to this node.
+    *
+    * This implementation prioritizes avoiding heap allocations.
+    * Since it does a linear search of the later events,
+    * this could be less efficient for very large sequences of events.
+    *
+    * @param events An ordered sequence of events to be de-duplicated.
+    * @param atTimeOverride If defined, the `EventTime` that should be combined with the single event;
+    *                       If empty, the an `EventTime` is obtained using `nextEventTime()`.
+    * @return a sequence of de-duplicated events wrapped in a `WithTime`.
+    *         This result is mutable, so it should not be allowed to escape this context.
+    */
+  def deduplicateEvents(events: Seq[NodeChangeEvent],
+                        atTimeOverride: Option[EventTime]
+                       ): Seq[WithTime] = {
+    require(atTimeOverride.isEmpty || events.length == 1, "Can only override EventTime for a single Event")
+
+    val r: ArrayBuffer[WithTime] = new ArrayBuffer[WithTime](events.length)
+
+    @tailrec
+    def accumulate(events: Seq[NodeChangeEvent] = events): Seq[WithTime] =
+      events match {
+        case event +: laterEvents =>
+
+          def laterEdgeEventExists(edge: HalfEdge): Boolean = laterEvents.exists {
+            case laterEvent: EdgeAdded => laterEvent.edge == edge
+            case laterEvent: EdgeRemoved => laterEvent.edge == edge
+            case _ => false
+          }
+
+          def laterPropertyEventExists(key: Symbol): Boolean = laterEvents.exists {
+            case laterEvent: PropertySet => laterEvent.key == key
+            case laterEvent: PropertyRemoved => laterEvent.key == key
+            case _ => false
+          }
+
+          def laterMergedHereExists(from: QuineId): Boolean = laterEvents.exists {
+            case laterEvent: MergedHere => laterEvent.from == from
+            case _ => false
+          }
+
+          val eventIsDuplicatedLater = event match {
+            case EdgeAdded(edge) => laterEdgeEventExists(edge)
+            case EdgeRemoved(edge) => laterEdgeEventExists(edge)
+            case PropertySet(key, _) => laterPropertyEventExists(key)
+            case PropertyRemoved(key, _) => laterPropertyEventExists(key)
+            case MergedIntoOther(_) => laterEvents.exists(_.isInstanceOf[MergedIntoOther])
+            case MergedHere(from) => laterMergedHereExists(from)
+          }
+
+          if (!eventIsDuplicatedLater && hasEffect(event))
+            r += WithTime(event, atTimeOverride.getOrElse(nextEventTime()))
+
+          accumulate(laterEvents)
+
+        case _ =>
+          r
```

Comment:
Note that since we cross-build 2.12 and 2.13 for this package: this code will not compile. `r: ArrayBuffer`, but in 2.13 (due to the collection overhaul) `ArrayBuffer`s are no longer `scala.Seq`s, since in 2.13, `scala.Seq[+A] = scala.collection.immutable.Seq[A]` rather than the more general (and mutable-permitting) `scala.collection.Seq[A]`